File: clients/swift/Tests/SynchroTests/ChangeTrackerTests.swift
import XCTest
import GRDB
@testable import Synchro

final class ChangeTrackerTests: XCTestCase {
    private func makeTestEnv() throws -> (SynchroDatabase, ChangeTracker, SchemaManager) {
        let tmpDir = NSTemporaryDirectory()
        let path = (tmpDir as NSString).appendingPathComponent("synchro_test_\(UUID().uuidString).sqlite")
        let db = try SynchroDatabase(path: path)
        let tracker = ChangeTracker(database: db)
        let manager = SchemaManager(database: db)

        let schema = SchemaResponse(
            schemaVersion: 1,
            schemaHash: protocolTestSchemaHash,
            serverTime: Date(),
            tables: [
                SchemaTable(
                    tableName: "orders",
                    pushPolicy: "owner_only",
                    updatedAtColumn: "updated_at",
                    deletedAtColumn: "deleted_at",
                    primaryKey: ["id"],
                    columns: [
                        SchemaColumn(name: "id", dbType: "uuid", logicalType: "string", nullable: false, isPrimaryKey: true),
                        SchemaColumn(name: "ship_address", dbType: "text", logicalType: "string", nullable: true, isPrimaryKey: false),
                        SchemaColumn(name: "user_id", dbType: "uuid", logicalType: "string", nullable: false, isPrimaryKey: false),
                        SchemaColumn(name: "created_at", dbType: "timestamp with time zone", logicalType: "datetime", nullable: false, isPrimaryKey: false),
                        SchemaColumn(name: "updated_at", dbType: "timestamp with time zone", logicalType: "datetime", nullable: false, isPrimaryKey: false),
                        SchemaColumn(name: "deleted_at", dbType: "timestamp with time zone", logicalType: "datetime", nullable: true, isPrimaryKey: false),
                    ]
                )
            ]
        )
        try manager.createSyncedTables(schema: schema)
        return (db, tracker, manager)
    }

    private func markOrderAsSynced(_ db: SynchroDatabase, tracker: ChangeTracker) throws {
        try db.writeTransaction { connection in
            try SynchroMeta.upsertRowVersion(
                connection,
                tableName: "orders",
                recordID: "w1",
                serverVersion: "server-version-1",
                rowChecksum: nil
            )
        }
        try tracker.clearAll()
    }

    private func makeTypedTestEnv() throws -> (SynchroDatabase, ChangeTracker, LocalSchemaTable) {
        let path = (NSTemporaryDirectory() as NSString).appendingPathComponent("synchro_typed_\(UUID().uuidString).sqlite")
        let db = try SynchroDatabase(path: path)
        let manager = SchemaManager(database: db)
        let table = SchemaTable(
            tableName: "typed_values",
            updatedAtColumn: "updated_at",
            deletedAtColumn: "deleted_at",
            primaryKey: ["id"],
            columns: [
                SchemaColumn(name: "id", logicalType: "string", nullable: false, isPrimaryKey: true),
                SchemaColumn(name: "int64_value", logicalType: "int64", nullable: false, isPrimaryKey: false),
                SchemaColumn(name: "int_value", logicalType: "int", nullable: false, isPrimaryKey: false),
                SchemaColumn(
                    name: "decimal_value",
                    logicalType: "decimal",
                    nullable: false,
                    precision: 6,
                    scale: 2,
                    isPrimaryKey: false
                ),
                SchemaColumn(name: "json_value", logicalType: "json", nullable: false, isPrimaryKey: false),
                SchemaColumn(name: "nullable_value", logicalType: "string", nullable: true, isPrimaryKey: false),
                SchemaColumn(name: "bool_value", logicalType: "boolean", nullable: false, isPrimaryKey: false),
                SchemaColumn(name: "float_value", logicalType: "float", nullable: false, isPrimaryKey: false),
                SchemaColumn(name: "blob_value", logicalType: "bytes", nullable: false, isPrimaryKey: false),
                SchemaColumn(name: "updated_at", logicalType: "datetime", nullable: false, isPrimaryKey: false),
                SchemaColumn(name: "deleted_at", logicalType: "datetime", nullable: true, isPrimaryKey: false),
            ]
        )
        try manager.createSyncedTables(
            schema: SchemaResponse(schemaVersion: 1, schemaHash: protocolTestSchemaHash, serverTime: Date(), tables: [table])
        )
        return (db, ChangeTracker(database: db), table)
    }

    func testInsertTriggerCreatesEntry() throws {
        let (db, tracker, _) = try makeTestEnv()

        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )

        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 1)
        XCTAssertEqual(pending[0].recordID, "w1")
        XCTAssertEqual(pending[0].tableName, "orders")
        XCTAssertEqual(pending[0].operation, "insert")
    }

    func testUpdateTriggerCreatesEntry() throws {
        let (db, tracker, _) = try makeTestEnv()

        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )
        try markOrderAsSynced(db, tracker: tracker)

        _ = try db.execute(
            "UPDATE orders SET ship_address = ? WHERE id = ?",
            params: ["456 Oak Ave", "w1"]
        )

        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 1)
        XCTAssertEqual(pending[0].operation, "update")
        XCTAssertNotNil(pending[0].baseUpdatedAt)
    }

    func testDeleteTriggerConvertsSoftDelete() throws {
        let (db, tracker, _) = try makeTestEnv()

        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )
        try markOrderAsSynced(db, tracker: tracker)

        // Hard DELETE should be intercepted by BEFORE DELETE trigger
        _ = try db.execute("DELETE FROM orders WHERE id = ?", params: ["w1"])

        // Record should still exist with deleted_at set
        let row = try db.queryOne("SELECT deleted_at FROM orders WHERE id = ?", params: ["w1"])
        XCTAssertNotNil(row)
        let deletedAt: String? = row?["deleted_at"]
        XCTAssertNotNil(deletedAt)

        // Pending queue should have a delete operation (from UPDATE trigger fired by BEFORE DELETE)
        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 1)
        XCTAssertEqual(pending[0].operation, "delete")
        XCTAssertTrue(pending[0].fieldValuesByID.isEmpty)
    }

    func testDeleteTriggerCapturesHardDeleteWhenTableHasNoSoftDeleteField() throws {
        let path = (NSTemporaryDirectory() as NSString)
            .appendingPathComponent("synchro_hard_delete_\(UUID().uuidString).sqlite")
        let db = try SynchroDatabase(path: path)
        let tracker = ChangeTracker(database: db)
        let manager = SchemaManager(database: db)
        let table = LocalSchemaTable(
            tableID: "hard_items",
            relationID: "public.hard_items",
            tableName: "hard_items",
            primaryKeyFieldID: "id",
            createdAtFieldID: nil,
            updatedAtFieldID: nil,
            deletedAtFieldID: nil,
            updatedAtColumn: "",
            deletedAtColumn: "",
            composition: .singleScope,
            primaryKey: ["id"],
            columns: [
                LocalSchemaColumn(
                    fieldID: "id", name: "id", logicalType: "string", nullable: false,
                    writable: false, precision: nil, scale: nil, sqliteDefaultSQL: nil, isPrimaryKey: true
                ),
                LocalSchemaColumn(
                    fieldID: "value", name: "value", logicalType: "string", nullable: false,
                    writable: true, precision: nil, scale: nil, sqliteDefaultSQL: nil, isPrimaryKey: false
                ),
            ]
        )
        try manager.reconcileLocalSchema(
            schemaVersion: 1,
            schemaHash: protocolTestSchemaHash,
            tables: [table]
        )
        _ = try db.execute(
            "INSERT INTO hard_items (id, value) VALUES (?, ?)",
            params: ["row-1", "value"]
        )
        try db.writeTransaction { connection in
            try SynchroMeta.upsertRowVersion(
                connection,
                tableName: table.tableName,
                recordID: "row-1",
                serverVersion: "server-version-1",
                rowChecksum: nil
            )
        }
        try tracker.clearAll()

        _ = try db.execute("DELETE FROM hard_items WHERE id = ?", params: ["row-1"])

        XCTAssertNil(try db.queryOne("SELECT id FROM hard_items WHERE id = ?", params: ["row-1"]))
        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 1)
        XCTAssertEqual(pending[0].operation, "delete")
        XCTAssertEqual(pending[0].baseUpdatedAt, "server-version-1")
        XCTAssertTrue(pending[0].fieldValuesByID.isEmpty)
    }

    func testDedupCreateThenUpdate() throws {
        let (db, tracker, _) = try makeTestEnv()

        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )
        _ = try db.execute(
            "UPDATE orders SET ship_address = ? WHERE id = ?",
            params: ["789 Updated Blvd", "w1"]
        )

        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 1)
        // insert + update normalizes to one insert.
        XCTAssertEqual(pending[0].operation, "insert")
        let sourceStates = try db.query(
            "SELECT mutation_id, lifecycle_state, normalized_mutation_id FROM _synchro_pending_changes WHERE table_name = 'orders' ORDER BY local_order",
            params: nil
        )
        XCTAssertEqual(sourceStates.count, 3)
        XCTAssertEqual(sourceStates[0]["lifecycle_state"] as String?, "superseded_before_send")
        XCTAssertEqual(sourceStates[1]["lifecycle_state"] as String?, "superseded_before_send")
        XCTAssertEqual(sourceStates[2]["lifecycle_state"] as String?, "unsealed")
        XCTAssertEqual(sourceStates[0]["normalized_mutation_id"] as String?, sourceStates[2]["mutation_id"] as String?)
    }

    func testDedupCreateThenDeleteRemovesEntry() throws {
        let (db, tracker, _) = try makeTestEnv()

        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )
        _ = try db.execute("DELETE FROM orders WHERE id = ?", params: ["w1"])

        // create + delete = removed entirely (never reached server)
        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 0)
        let cancelled = try db.query(
            "SELECT lifecycle_state, normalized_mutation_id FROM _synchro_pending_changes WHERE table_name = 'orders' ORDER BY local_order",
            params: nil
        )
        XCTAssertEqual(cancelled.count, 2)
        XCTAssertTrue(cancelled.allSatisfy { ($0["lifecycle_state"] as String?) == "cancelled_before_send" })
        XCTAssertNotNil(cancelled[0]["normalized_mutation_id"] as String?)
        XCTAssertEqual(cancelled[0]["normalized_mutation_id"] as String?, cancelled[1]["normalized_mutation_id"] as String?)
    }

    func testSyncLockPreventsTracking() throws {
        let (db, tracker, _) = try makeTestEnv()

        try db.writeTransaction { dbConn in
            try SynchroMeta.setSyncLock(dbConn, locked: true)
        }

        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )

        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 0)

        try db.writeTransaction { dbConn in
            try SynchroMeta.setSyncLock(dbConn, locked: false)
        }
    }

    func testClearAll() throws {
        let (db, tracker, _) = try makeTestEnv()

        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )

        XCTAssertTrue(try tracker.hasPendingChanges())
        try tracker.clearAll()
        XCTAssertFalse(try tracker.hasPendingChanges())
    }

    func testDedupUpdateThenDelete() throws {
        let (db, tracker, _) = try makeTestEnv()

        // Insert and clear (simulate already-pushed create)
        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )
        try markOrderAsSynced(db, tracker: tracker)

        // Update then delete — should dedup to "delete"
        _ = try db.execute("UPDATE orders SET ship_address = ? WHERE id = ?", params: ["555 Renamed Dr", "w1"])
        _ = try db.execute("DELETE FROM orders WHERE id = ?", params: ["w1"])

        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 1)
        XCTAssertEqual(pending[0].operation, "delete")
        // base_updated_at should be preserved from the update (not nil — record existed on server)
        XCTAssertNotNil(pending[0].baseUpdatedAt)
    }

    func testDedupUpdateThenUpdate() throws {
        let (db, tracker, _) = try makeTestEnv()

        // Insert and clear (simulate already-pushed create)
        _ = try db.execute(
            "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            params: ["w1", "123 Main St", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
        )
        try markOrderAsSynced(db, tracker: tracker)

        // Two sequential updates — should dedup to single "update"
        _ = try db.execute("UPDATE orders SET ship_address = ? WHERE id = ?", params: ["100 First Ave", "w1"])
        _ = try db.execute("UPDATE orders SET ship_address = ? WHERE id = ?", params: ["200 Second Blvd", "w1"])

        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 1)
        XCTAssertEqual(pending[0].operation, "update")
        // base_updated_at should be from the original record (before first update)
        XCTAssertNotNil(pending[0].baseUpdatedAt)
    }

    func testUnhydratablePendingChangeFailsWithoutDeletingIntent() throws {
        let (db, tracker, _) = try makeTestEnv()
        try db.writeTransaction { connection in
            try connection.execute(
                sql: """
                    INSERT INTO _synchro_pending_changes
                        (mutation_id, capture_uuid, table_id, table_name, record_id, pk_field_id, pk_logical_type,
                         operation, authored_schema_version, authored_schema_hash, client_version, lifecycle_state,
                         source_kind, created_at, updated_at)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'unsealed', 'test', ?, ?)
                    """,
                arguments: [
                    UUID().uuidString.lowercased(), UUID().uuidString.lowercased(), "orders", "orders", "missing-1", "id", "string",
                    "insert", 1, protocolTestSchemaHash, "2026-01-01T00:00:00.000000Z", "2026-01-01T00:00:00.000000Z", "2026-01-01T00:00:00.000000Z",
                ]
            )
        }

        let pending = try tracker.pendingChanges()
        XCTAssertThrowsError(try tracker.hydratePendingForPush(pending: pending, syncedTables: [])) { error in
            guard case SynchroError.invalidResponse = error else {
                return XCTFail("expected explicit hydration failure, got \(error)")
            }
        }
        XCTAssertEqual(try tracker.pendingChangeCount(), 1)
    }

    func testTypedCaptureRetainsImmutableValuesAndWireForms() throws {
        let (db, tracker, table) = try makeTypedTestEnv()
        let blob = Data([0x00, 0x01, 0x02])
        _ = try db.execute(
            """
            INSERT INTO typed_values
                (id, int64_value, int_value, decimal_value, json_value, nullable_value, bool_value, float_value, blob_value, updated_at)
            VALUES (?, ?, ?, ?, ?, NULL, ?, ?, ?, ?)
            """,
            params: ["t1", "9223372036854775807", 7, "12.3400", "{\"b\":2,\"a\":1}", 1, 3.5, blob, "2026-01-01T00:00:00Z"]
        )

        let pending = try tracker.pendingChanges()
        XCTAssertEqual(pending.count, 1)
        let hydrated = try tracker.hydratePendingForPush(pending: pending, syncedTables: [table])
        let fields = try XCTUnwrap(hydrated.first?.fieldValuesByID)
        XCTAssertEqual(fields["int64_value"]?.wireValue, AnyCodable("9223372036854775807"))
        XCTAssertEqual(fields["int_value"]?.wireValue, AnyCodable(Int64(7)))
        XCTAssertEqual(fields["decimal_value"]?.wireValue, AnyCodable("12.3400"))
        XCTAssertEqual(fields["json_value"]?.wireValue, AnyCodable("{\"b\":2,\"a\":1}"))
        XCTAssertEqual(fields["nullable_value"]?.wireValue, AnyCodable(NSNull()))
        XCTAssertEqual(fields["bool_value"]?.wireValue, AnyCodable(true))
        XCTAssertEqual(fields["float_value"]?.wireValue, AnyCodable(3.5))
        XCTAssertEqual(fields["blob_value"]?.wireValue, AnyCodable("AAEC"))

        try db.writeSyncLockedTransaction { connection in
            try connection.execute(
                sql: "UPDATE typed_values SET int64_value = ?, decimal_value = ?, json_value = ?, blob_value = ? WHERE id = ?",
                arguments: ["1", "changed", "changed", Data([0xff]), "t1"]
            )
        }
        let afterRowMutation = try tracker.pendingChanges()
        let retained = try tracker.hydratePendingForPush(pending: afterRowMutation, syncedTables: [table])
        let retainedFields = try XCTUnwrap(retained.first?.fieldValuesByID)
        XCTAssertEqual(retainedFields["int64_value"]?.wireValue, AnyCodable("9223372036854775807"))
        XCTAssertEqual(retainedFields["decimal_value"]?.wireValue, AnyCodable("12.3400"))
        XCTAssertEqual(retainedFields["blob_value"]?.wireValue, AnyCodable("AAEC"))
    }

    func testSchemaArchiveMismatchRollsBackApplicationWriteAndCapture() throws {
        let (db, tracker, _) = try makeTestEnv()
        try db.writeTransaction { connection in
            try connection.execute(sql: "DELETE FROM _synchro_schema_archive")
        }
        XCTAssertThrowsError(
            try db.execute(
                "INSERT INTO orders (id, ship_address, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
                params: ["archive-mismatch", "blocked", "u1", "2026-01-01T00:00:00Z", "2026-01-01T00:00:00Z"]
            )
        )
        XCTAssertNil(try db.queryOne("SELECT id FROM orders WHERE id = 'archive-mismatch'", params: nil))
        XCTAssertEqual(try tracker.pendingChangeCount(), 0)
    }
}

File: clients/kotlin/synchro/src/test/kotlin/com/trainstar/synchro/ChangeTrackerTests.kt
package com.trainstar.synchro

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import java.util.UUID
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [28])
class ChangeTrackerTests {
    private val databases = TestDatabaseTracker()

    private val table = SchemaTable(
        tableName = "orders",
        updatedAtColumn = "updated_at",
        deletedAtColumn = "deleted_at",
        primaryKey = listOf("id"),
        columns = listOf(
            SchemaColumn("id", logicalType = "string", nullable = false, isPrimaryKey = true),
            SchemaColumn("title", logicalType = "string"),
            SchemaColumn("enabled", logicalType = "boolean"),
            SchemaColumn("quantity", logicalType = "int"),
            SchemaColumn("large_quantity", logicalType = "int64"),
            SchemaColumn("amount", logicalType = "decimal", precision = 6, scale = 2),
            SchemaColumn("document", logicalType = "json"),
            SchemaColumn("score", logicalType = "float"),
            SchemaColumn("payload", logicalType = "bytes"),
            SchemaColumn("updated_at", logicalType = "datetime", nullable = false),
            SchemaColumn("deleted_at", logicalType = "datetime"),
        ),
    )

    private fun environment(): Pair<SynchroDatabase, ChangeTracker> {
        val context = ApplicationProvider.getApplicationContext<Context>()
        val database = databases.create(context)
        SchemaManager(database).createSyncedTables(
            SchemaResponse(1, PROTOCOL_TEST_SCHEMA_HASH, "2026-01-01T00:00:00.000000Z", listOf(table)),
        )
        return database to ChangeTracker(database)
    }

    @After
    fun tearDown() = databases.closeAll()

    @Test
    fun captureAppendsImmutableTypedValues() {
        val (database, tracker) = environment()
        database.execute(
            """
            INSERT INTO orders (id, title, enabled, quantity, large_quantity, amount, document, score, payload, updated_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """.trimIndent(),
            arrayOf(
                "o1", null, 1L, 7L, 9_007_199_254_740_991L, "123.45", "{\"a\":1,\"b\":true}",
                2.5, byteArrayOf(1, 2, 3), "2026-01-01T00:00:00.000000Z",
            ),
        )
        database.execute("UPDATE orders SET title = ?, score = ? WHERE id = ?", arrayOf("after", 3.5, "o1"))
        database.execute("UPDATE orders SET title = ? WHERE id = ?", arrayOf("later", "o1"))

        val ledger = database.query(
            "SELECT mutation_id, local_order, operation, depends_on_mutation_id FROM _synchro_pending_changes ORDER BY local_order",
        )
        assertEquals(3, ledger.size)
        assertEquals(listOf("insert", "update", "update"), ledger.map { it.getValue("operation") })
        ledger.forEach { assertEquals(it.getValue("mutation_id"), UUID.fromString(it.getValue("mutation_id") as String).toString()) }
        assertNotNull(ledger[1]["depends_on_mutation_id"])
        assertEquals(ledger[1]["mutation_id"], ledger[2]["depends_on_mutation_id"])

        val insertID = ledger.first().getValue("mutation_id") as String
        val updateID = ledger[1].getValue("mutation_id") as String
        val insertValues = database.query(
            "SELECT field_id, logical_type, value_kind, value_integer, value_real, value_blob FROM _synchro_mutation_values WHERE mutation_id = ? ORDER BY field_id",
            arrayOf(insertID),
        )
        assertEquals(8, insertValues.size)
        assertEquals("null", insertValues.single { it["field_id"] == "title" }["value_kind"])
        assertEquals("boolean", insertValues.single { it["field_id"] == "enabled" }["value_kind"])
        assertEquals("integer", insertValues.single { it["field_id"] == "quantity" }["value_kind"])
        assertEquals("integer", insertValues.single { it["field_id"] == "large_quantity" }["value_kind"])
        assertEquals("text", insertValues.single { it["field_id"] == "amount" }["value_kind"])
        assertEquals("text", insertValues.single { it["field_id"] == "document" }["value_kind"])
        assertEquals("real", insertValues.single { it["field_id"] == "score" }["value_kind"])
        assertEquals("blob", insertValues.single { it["field_id"] == "payload" }["value_kind"])

        val updateFields = database.query(
            "SELECT field_id FROM _synchro_mutation_values WHERE mutation_id = ? ORDER BY field_id",
            arrayOf(updateID),
        ).map { it.getValue("field_id") }
        assertEquals(listOf("score", "title"), updateFields)
        val hydrated = tracker.hydratePendingForPush(listOf(tracker.pendingChanges().first()), listOf(table.localSchema)).single()
        assertEquals(AnyCodable(7L), hydrated.data?.get("quantity"))
        assertEquals(AnyCodable("9007199254740991"), hydrated.data?.get("large_quantity"))
        assertEquals(AnyCodable("123.45"), hydrated.data?.get("amount"))
        assertEquals(AnyCodable("{\"a\":1,\"b\":true}"), hydrated.data?.get("document"))
        assertEquals(AnyCodable(true), hydrated.data?.get("enabled"))
        assertEquals(AnyCodable(2.5), hydrated.data?.get("score"))
        assertEquals(AnyCodable("AQID"), hydrated.data?.get("payload"))
        assertEquals(AnyCodable(null), hydrated.data?.get("title"))
        assertEquals(3, tracker.pendingChangeCount())
    }

    @Test
    fun deleteAppendsAColumnFreeIntentWithoutCoalescingSources() {
        val (database, _) = environment()
        database.execute(
            "INSERT INTO orders (id, title, updated_at) VALUES (?, ?, ?)",
            arrayOf("o1", "first", "2026-01-01T00:00:00.000000Z"),
        )
        database.execute("UPDATE orders SET title = ? WHERE id = ?", arrayOf("second", "o1"))
        database.execute("DELETE FROM orders WHERE id = ?", arrayOf("o1"))

        val changes = database.query("SELECT mutation_id, operation FROM _synchro_pending_changes ORDER BY local_order")
        assertEquals(listOf("insert", "update", "delete"), changes.map { it.getValue("operation") })
        val deleteID = changes.last().getValue("mutation_id") as String
        assertTrue(database.query("SELECT field_id FROM _synchro_mutation_values WHERE mutation_id = ?", arrayOf(deleteID)).isEmpty())
        assertNotNull(database.queryOne("SELECT deleted_at FROM orders WHERE id = ?", arrayOf("o1"))?.get("deleted_at"))
    }

    @Test
    fun syncLockAndMissingArchivePreventCaptureAtomically() {
        val (database, tracker) = environment()
        database.writeTransaction { SynchroMeta.setSyncLock(it, true) }
        database.execute(
            "INSERT INTO orders (id, title, updated_at) VALUES (?, ?, ?)",
            arrayOf("locked", "no intent", "2026-01-01T00:00:00.000000Z"),
        )
        assertFalse(tracker.hasPendingChanges())

        database.writeTransaction {
            SynchroMeta.setSyncLock(it, false)
            it.execSQL("DELETE FROM _synchro_schema_archives")
        }
        assertTrue(
            runCatching {
                database.execute(
                    "INSERT INTO orders (id, title, updated_at) VALUES (?, ?, ?)",
                    arrayOf("blocked", "no row", "2026-01-01T00:00:00.000000Z"),
                )
            }.isFailure,
        )
        assertTrue(database.query("SELECT id FROM orders WHERE id = 'blocked'").isEmpty())
        assertEquals(0L, database.queryOne("SELECT COUNT(*) AS count FROM _synchro_pending_changes")?.get("count"))
    }
}
